### PR TITLE
fix data inconsistent when runc update in systemd driven cgroup

### LINKED ISSUE #2287: runc --systemd-cgroup update does not update systemd scope
A repro

1. Prepare a bundle and a spec. Amend the config.json with
```json
        "pids": {
                "limit": 55
        }
```
in the `linux.resources` section, and a long-lived process such as sleep.

2. Start the container:

```console
# runc --systemd-cgroup run -d keke
```

3. Check its pids limit in cgroups scope, as well as in cgroupfs:

```console
# systemctl show runc-keke.scope | grep 'Task'
TasksCurrent=2
TasksAccounting=yes
TasksMax=55

# cat /sys/fs/cgroup/system.slice/runc-keke.scope/pids.max 
55
```

4. Update the limit:
```console
# runc --systemd-cgroup update keke --pids-limit 66
```

5. Check again:

```console
# systemctl show runc-keke.scope | grep 'Task'
TasksCurrent=2
TasksAccounting=yes
TasksMax=55

# cat /sys/fs/cgroup/system.slice/runc-keke.scope/pids.max 
66
```

As we can see, the kernel and the systemd is now in disagreement about the limit.

6. Check that the limit can in principal be updated:

```console
# systemctl show runc-keke.scope | grep 'Task'
TasksCurrent=2
TasksAccounting=yes
TasksMax=77
DropInPaths=/run/systemd/transient/runc-keke.scope.d/50-TasksMax.conf
# cat /sys/fs/cgroup/system.slice/runc-keke.scope/pids.max 
77
```

In the above example, we set/check/update pids.limit, but all the other resources behave in a similar way.

The cause of this is `Set()` method of `systemd.UnifiedManager`, which only uses fsManager to set the updated values.

### PR DESCRIPTION
**Pull request by lifubang**
fix #2287 
`runc --systemd-cgroup update does not update systemd scope`

As pointed out by @kolyshkin `The cause of this is Set() method of systemd.UnifiedManager, which only uses fsManager to set the updated values.`
So when we use systemd driven cgroup v2, we should use `SetUnitProperties` method to update container resource constraints.

Signed-off-by: lifubang <lifubang@acmcoder.com>

### CODE REVIEW

**Comment by lifubang** [comment]
If it makes sense, the same changes can be added to cgroup v1.

File: libcontainer/cgroups/systemd/v2.go [hunk 1]
```diff
@@ -263,11 +274,23 @@ func (m *unifiedManager) GetStats() (*cgroups.Stats, error) {
...
-	fsMgr, err := m.fsManager()
-	if err != nil {
+	var (
+		c        = m.cgroups
+		unitName = getUnitName(c)
+	)
+
+	if properties, err := genResourcesProperties(c); err != nil {
 		return err
+	} else {
```
Comment (kolyshkin) [inline libcontainer/cgroups/systemd/v2.go:284]: no need for "else" here since you have `return` above.

File: libcontainer/cgroups/systemd/v2.go [hunk 2]
```diff
@@ -82,55 +135,13 @@ func (m *unifiedManager) Apply(pid int) error {
...
-	if c.Resources.Memory != 0 {
-		properties = append(properties,
-			newProp("MemoryMax", uint64(c.Resources.Memory)))
-	}
-
-	swap, err := cgroups.ConvertMemorySwapToCgroupV2Value(c.Resources.MemorySwap, c.Resources.Memory)
-	if err != nil {
+	if resourcesProperties, err := genResourcesProperties(c); err != nil {
 		return err
+	} else {
```
Comment (kolyshkin) [inline libcontainer/cgroups/systemd/v2.go:140]: no need for `else` here

File: libcontainer/cgroups/systemd/v2.go [hunk 3]
```diff
@@ -263,11 +274,23 @@ func (m *unifiedManager) GetStats() (*cgroups.Stats, error) {
...
+		dbusConnection, err := getDbusConnection()
+		if err != nil {
+			return err
+		}
+		if err := dbusConnection.SetUnitProperties(unitName, true, properties...); err != nil {
+			return err
+		}
+		return nil
 	}
-	return fsMgr.Set(container)
```
Comment (kolyshkin) [inline libcontainer/cgroups/systemd/v2.go:270]: some resources are not controlled by systemd so we still need to call this I'm afraid.

Comment (kolyshkin) [reply libcontainer/cgroups/systemd/v2.go:270]: and this is why CI is failing (systemd does not control cpuset)

Comment (lifubang) [reply libcontainer/cgroups/systemd/v2.go]: Force pushed now.
I don't know whether we should change to only write `cpuset` cgroup files or not?

**Comment by lifubang** [comment]
> but as you say it also applies to cgroup v1, so it would be nice if you could fix that at the same time!

OK, I'll fix and test it in cgroup v1.

File: libcontainer/cgroups/systemd/v2.go (hunk 3 above)
Comment (odinuge) [reply libcontainer/cgroups/systemd/v2.go]: I think the best is to keep it like it is now (with writing all of them), and then we can look at it in another PR/issue. This also applies to other resources not set via systemd like hugetlb. 

More info about it here: https://systemd.io/CGROUP_DELEGATION/

**Comment by kolyshkin** [comment]
LGTM

**Comment by kolyshkin** [comment]
@lifubang maybe we can add some simple tests, based on what I described in #2287.

**Comment by lifubang** [comment]
> maybe we can add some simple tests

OK, I'll do it later.

**Comment by lifubang** [comment]
Three tests has been added.

**Comment by AkihiroSuda** [comment]
LGTM

**Comment by kolyshkin** [comment]
LGTM

**Pull request merged**